clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -3070,4 +3072,69 @@ void handleFailure(Throwable throwable) {
 
         return new ElectLeadersResult(electionFuture);
     }
+
+    @Override
+    public MembershipChangeResult removeMemberFromGroup(String groupId,
+                                                        RemoveMemberFromGroupOptions options) {
+        final long startFindCoordinatorMs = time.milliseconds();
+        final long deadline = calcDeadlineMs(startFindCoordinatorMs, options.timeoutMs());
+
+        KafkaFutureImpl<RemoveMemberFromGroupResult> future = new KafkaFutureImpl<>();
+
+        ConsumerGroupOperationContext<RemoveMemberFromGroupResult, RemoveMemberFromGroupOptions> context =
+            new ConsumerGroupOperationContext<>(groupId, options, deadline, future);
+
+        Call findCoordinatorCall = getFindCoordinatorCall(context,
+            () -> KafkaAdminClient.this.getRemoveMembersFromGroupCall(context));
+        runnable.call(findCoordinatorCall, startFindCoordinatorMs);
+
+        return new MembershipChangeResult(future);
+    }
+
+    private Call getRemoveMembersFromGroupCall(ConsumerGroupOperationContext
+                                                   <RemoveMemberFromGroupResult, RemoveMemberFromGroupOptions> context) {
+        return new Call("removeMembersFromGroup",
+                        context.getDeadline(),
+                        new ConstantNodeIdProvider(context.getNode().get().id())) {
+            @Override
+            AbstractRequest.Builder createRequest(int timeoutMs) {
+                return new LeaveGroupRequest.Builder(context.getGroupId(),
+                                                     context.getOptions().getMembers());
+            }
+
+            @Override
+            void handleResponse(AbstractResponse abstractResponse) {
+                final LeaveGroupResponse response = (LeaveGroupResponse) abstractResponse;
+
+                // If coordinator changed since we fetched it, retry
+                if (context.hasCoordinatorMoved(response)) {
+                    rescheduleTask(context, () -> getRemoveMembersFromGroupCall(context));
+                    return;
+                }
+
+                Errors error = response.error();
+
+                // If this is a known top level error, we will throw exception directly. This is
+                // because no member should have been removed successfully.
+                // For transit coordinator error, the whole process will retry.
+                if (error == Errors.COORDINATOR_LOAD_IN_PROGRESS
+                        || error == Errors.COORDINATOR_NOT_AVAILABLE
+                        || error == Errors.GROUP_AUTHORIZATION_FAILED
+                        || error == Errors.NOT_COORDINATOR) {
+                    throw error.exception();
+                }
```

Comment:
Throwing an exception here would just cause a `caller.fail`, and then caused a `handleFailure` instead.

I think it's better just setting the exception in the future directly.

Reply:
Also, we should handle all errors as well so that when constructing the result we know it's successful and hence we do not need to set any errors below.